### PR TITLE
[image_picker] add requestFullMetadata for iOS (optional permissions) - iOS changes

### PR DESCRIPTION
**Pull request by PiotrMitkowski**
## Description

iOS changes for https://github.com/flutter/plugins/pull/5915

## Related issues
flutter/flutter#65995

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by PiotrMitkowski** [comment]
@stuartmorgan @cyanglaz I moved iOS specific changes to a separate branch, as suggested.

File: packages/image_picker/image_picker_ios/lib/image_picker_ios.dart [hunk 1]
```diff
@@ -51,14 +51,28 @@ class ImagePickerIOS extends ImagePickerPlatform {
...
+        maxHeight: maxHeight,
+        imageQuality: imageQuality,
+        preferredCameraDevice: preferredCameraDevice,
+      ),
     );
     return path != null ? PickedFile(path) : null;
   }
 
+  @override
+  Future<XFile?> getImageFromSource({
```
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/lib/image_picker_ios.dart:65]: Maybe we should deprecate `pickImage` as `getImageFromSource` is the new method we want users to start using instead.

**Review by cyanglaz — CHANGES_REQUESTED** [review]

File: packages/image_picker/image_picker_ios/lib/image_picker_ios.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/image_picker/image_picker_ios/lib/image_picker_ios.dart:65]: Actually deprecating platform interface methods is a multi-step process; see discussion of deprecation in [this section of the docs](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-platform-interface-method-parameters)

File: packages/image_picker/image_picker_ios/example/pubspec.yaml [hunk 2]
```diff
@@ -16,7 +16,7 @@ dependencies:
     # The example app is bundled with the plugin so we use a path dependency on
     # the parent directory to use the current plugin's version.
     path: ../
-  image_picker_platform_interface: ^2.3.0
+  image_picker_platform_interface: ^2.5.0
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/image_picker/image_picker_ios/example/pubspec.yaml:19]: 2.4.0 was the broken version that I reverted in 2.4.1

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m [hunk 3]
```diff
@@ -119,7 +119,12 @@ - (void)launchPHPickerWithContext:(nonnull FLTImagePickerMethodCallContext *)con
   _pickerViewController.presentationController.delegate = self;
   self.callContext = context;
 
-  [self checkPhotoAuthorizationForAccessLevel];
+  BOOL requestFullMetadata = context.requestFullMetadata;
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:122]: Nit: there's no value to this local variable given how short and readable the right hand side is; you can just inline `context.requestFullMetadata` into the `if` condition.

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m [hunk 4]
```diff
@@ -129,14 +134,24 @@ - (void)launchUIImagePickerWithSource:(nonnull FLTSourceSpecification *)source
   imagePickerController.delegate = self;
   imagePickerController.mediaTypes = @[ (NSString *)kUTTypeImage ];
   self.callContext = context;
+  BOOL requestFullMetadata = context.requestFullMetadata;
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:137]: Same here. Especially since this is only used once, and conditionally, so doesn't make sense to define at such a high level.

File: packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m [hunk 5]
```diff
@@ -177,6 +179,27 @@ - (void)testPickMultiImageShouldUseUIImagePickerControllerOnPreiOS14 {
             [mockUIImagePicker setSourceType:UIImagePickerControllerSourceTypePhotoLibrary]);
 }
 
+- (void)testPickImageWithoutFullMetadataPreiOS14 {
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m:182]: Why is this test pre-14?

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m [hunk 6]
```diff
@@ -227,14 +244,19 @@ - (void)pickVideoWithSource:(nonnull FLTSourceSpecification *)source
   }
 
   self.callContext = context;
+  BOOL requestFullMetadata = context.requestFullMetadata;
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:247]: Same here; just inline the expression.

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m [hunk 7]
```diff
@@ -553,8 +575,13 @@ - (void)imagePickerController:(UIImagePickerController *)picker
     NSNumber *maxHeight = self.callContext.maxSize.height;
     NSNumber *imageQuality = self.callContext.imageQuality;
     NSNumber *desiredImageQuality = [self getDesiredImageQuality:imageQuality];
+    BOOL requestFullMetadata = _callContext.requestFullMetadata;
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:578]: Same.

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m [hunk 8]
```diff
@@ -553,8 +575,13 @@ - (void)imagePickerController:(UIImagePickerController *)picker
...
 
-    PHAsset *originalAsset = [FLTImagePickerPhotoAssetUtil getAssetFromImagePickerInfo:info];
+    PHAsset *originalAsset;
+    if (requestFullMetadata) {
+      // Full metadata are available only in PHAsset, which requires gallery permission
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:582]: Nit: missing period.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Looks good with some minor nits

File: packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m (hunk 5 above)
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m:182]: I realized that it was a wrong test case, not fully covering the new flag's use case and vulnerable to false positives. I've changed it so it now verifies that there was no check for gallery authorization status.

**Comment by jmagman** [comment]
Friendly bump, looks like the nits were addressed.

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m [hunk 9]
```diff
@@ -234,7 +249,11 @@ - (void)pickVideoWithSource:(nonnull FLTSourceSpecification *)source
                                              camera:[self cameraDeviceForSource:source]];
       break;
     case FLTSourceTypeGallery:
-      [self checkPhotoAuthorizationWithImagePicker:imagePickerController];
+      if (context.requestFullMetadata) {
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:252]: Doesn't this need the same `@available` check and fallback?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Sorry, one more issue I missed last time.

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m (hunk 9 above)
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:252]: I just realized that this change for picking a video can't be used with the current platform interface. It doesn't allow passing `requestFullMetadata` for both multi-image and videos. I can either improve the interface or remove this change for video picking (depending on what you think is better).

Comment (stuartmorgan) [reply packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:252]: I don't see any reason we wouldn't want support for this in the multi-image case, so updating that interface makes sense.

Looking at the result callback, is this change even relevant for video? It looks like the parameter is only used in a non-video codepath.

**Review by cyanglaz — APPROVED** [review]
LGTM

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m (hunk 9 above)
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:252]: Okay, I'll update the interface and create a PR soon. I'll also remove usage of `requestFullMetadata` for the video path, since it it isn't used, as you noticed.

**Comment by PiotrMitkowski** [comment]
@cyanglaz Sure, I got stuck on waiting for review of this one: https://github.com/flutter/plugins/pull/5914, which blocks further work on adding `requestFullMetadata` flag. I wanted to avoid too much pinging, but I guess it's a good time to post it on Discord :)

**Comment by cyanglaz** [comment]
I will take a look at https://github.com/flutter/plugins/pull/5914

**Comment by PiotrMitkowski** [comment]
@stuartmorgan @cyanglaz It's ready for your review after implementing optional permissions for multi image picking

File: packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m [hunk 10]
```diff
@@ -170,13 +172,49 @@ - (void)testPickMultiImageShouldUseUIImagePickerControllerOnPreiOS14 {
...
                          }];
   OCMVerify(times(1),
             [mockUIImagePicker setSourceType:UIImagePickerControllerSourceTypePhotoLibrary]);
 }
 
+- (void)testPickImageWithoutFullMetadata API_AVAILABLE(ios(11)) {
+  id mockUIImagePicker = OCMClassMock([UIImagePickerController class]);
+  id photoLibrary = OCMClassMock([PHPhotoLibrary class]);
+
+  FLTImagePickerPlugin *plugin = [FLTImagePickerPlugin new];
```
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m:187]: nits: do not use +new
https://google.github.io/styleguide/objcguide.html#do-not-use-new

File: packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m [hunk 11]
```diff
@@ -170,13 +172,49 @@ - (void)testPickMultiImageShouldUseUIImagePickerControllerOnPreiOS14 {
...
+                   }];
+
+  OCMVerify(times(0), [photoLibrary authorizationStatus]);
+}
+
+- (void)testPickMultiImageWithoutFullMetadata API_AVAILABLE(ios(11)) {
+  id mockUIImagePicker = OCMClassMock([UIImagePickerController class]);
+  id photoLibrary = OCMClassMock([PHPhotoLibrary class]);
+
+  FLTImagePickerPlugin *plugin = [FLTImagePickerPlugin new];
```
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m:205]: ditto

**Comment by PiotrMitkowski** [comment]
@cyanglaz I've fixed the code based on your latest remarks. Right now, I can't see an option to request your another review, so I'm pinging through this comment :)

**Review by cyanglaz — APPROVED** [review]
LGTM
cc @stuartmorgan

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m (hunk 9 above)
Comment (stuartmorgan) [reply packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:254]: Looks like this is still missing the iOS 11 `@available` check.

File: packages/image_picker/image_picker_ios/lib/image_picker_ios.dart [hunk 12]
```diff
@@ -2,12 +2,9 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'dart:async';
-
+import 'package:image_picker_ios/src/messages.g.dart';
 import 'package:image_picker_platform_interface/image_picker_platform_interface.dart';
 
-import 'src/messages.g.dart';
```
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/lib/image_picker_ios.dart:9]: Why did you change this to a package-style import within the same package?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
A couple of small things, but otherwise looks good.

File: packages/image_picker/image_picker_ios/lib/image_picker_ios.dart (hunk 12 above)
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker_ios/lib/image_picker_ios.dart:9]: I was checking if it could fix an issue with imports related to Pigeon, but then I've found a GitHub issue about it and forgot to revert this change. I'll do it now.

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m (hunk 9 above)
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m]: I've removed usage of `requestFullMetadata` from here since it isn't supported in the video picking.

**Review by stuartmorgan — APPROVED** [review]
LGTM!

**Comment by idish** [comment]
Thanks for the great work!
Is there any schedule for the 0.8.6 version to be published? Really looking forward to it.

**Comment by jmagman** [comment]
It's already available, note it's `image_picker_ios` https://pub.dev/packages/image_picker_ios/changelog#086

**Comment by stuartmorgan** [comment]
The PR you'll want to follow to know when you can use this feature as a plugin client is https://github.com/flutter/plugins/pull/5915
